activesupport/lib/active_support/cache/coder.rb

```diff
@@ -0,0 +1,142 @@
+# frozen_string_literal: true
+
+require_relative "entry"
+
+module ActiveSupport
+  module Cache
+    class Coder # :nodoc:
+      def initialize(serializer, compressor, legacy_serializer: false)
+        @serializer = serializer
+        @compressor = compressor
+        @legacy_serializer = legacy_serializer
+      end
+
+      def dump(entry)
+        return @serializer.dump(entry) if @legacy_serializer
+
+        dump_compressed(entry, Float::INFINITY)
+      end
+
+      def dump_compressed(entry, threshold)
+        return @serializer.dump_compressed(entry, threshold) if @legacy_serializer
+
+        # If value is a string with a supported encoding, use it as the payload
+        # instead of passing it through the serializer.
+        if type = type_for_string(entry.value)
+          payload = binary_compatible_string(entry.value)
+        else
+          type = OBJECT_DUMP_TYPE
+          payload = serialize(entry.value)
+        end
+
+        # Compress payload if it meets the size threshold. (Note that `type < 0`
+        # signifies a compressed payload with `type.abs` as its actual type.)
+        if compressed = try_compress(payload, threshold)
+          payload = compressed
+          type = -type
+        end
+
+        expires_at = entry.expires_at || -1.0
+
+        version = dump_version(entry.version)
+        version_length = version&.bytesize || -1
+
+        packed = [SIGNATURE, type, expires_at, version_length].pack(PACKED_TEMPLATE)
```

Comment:
```suggestion
        packed = SIGNATURE.b
        packed << [type, expires_at, version_length].pack(PACKED_TEMPLATE)
```

A bit of a nitpick, but I think it would be a bit clearer if the signature wasn't part of the template on `dump`.

Reply:
I've made this change, but I want to point out that it creates disparity between `PACKED_TEMPLATE` and the other `PACKED_*_TEMPLATE` strings, since the latter must still account for the signature.

Reply:
Right, I don't think it's a big deal. As long as it's covered by test. All these templates are quite cryptic in the first place 😄